src/jupyterhub_config/custom_spawner.py

```diff
@@ -126,4 +133,38 @@ def _ensure_user_jupyter_directory(self, user_dir: Path):
 
         self.environment['JUPYTER_CONFIG_DIR'] = str(jupyter_dir)
         self.environment['JUPYTER_RUNTIME_DIR'] = str(jupyter_runtime_dir)
-        self.environment['JUPYTER_DATA_DIR'] = str(juputer_data_dir)
\ No newline at end of file
+        self.environment['JUPYTER_DATA_DIR'] = str(juputer_data_dir)
+
+    def _ensure_virtual_environment(self, user_env_dir: Path):
+        """
+        Ensure the user's virtual environment exists. If it does not exist, it is
+        created with the system site-packages included.
+        """
+        if not user_env_dir.exists():
+            user_env_dir.mkdir(parents=True)
+            self.log.info(f'Creating virtual environment for {self.user.name}')
+            try:
+                # Create a virtual environment with system site-packages access
+                subprocess.run(['python3', '-m', 'venv', str(user_env_dir), '--system-site-packages'],
+                               check=True)
+            except subprocess.CalledProcessError as e:
+                self.log.error(f'Failed to create virtual environment: {e}')
+                raise
+        else:
+            self.log.info(f'Reusing virtual environment for {self.user.name}')
+
+    def _configure_environment(self, user_dir: Path, user_env_dir: Path, username: str):
+        """
+        Configure the environment variables for the user's session, including
+        the PATH and PYTHONPATH to use the virtual environment.
+        """
+        self.environment.update({key: value for key, value in os.environ.items() if key not in self.environment})
```

Comment:
This is the jupyterhub environment I guess? Is it not already populated with the os environment?

Reply:
It's the spawner environment variable (user's individual environment). I think it's good to copy os environment to individual user's. Maybe we need a white list in the future to hide some os env from users.

Reply:
It's the spawner environment variable (user's individual environment). I think it's good to copy os environment to individual user's. Maybe we need a white list in the future to hide some os env from users.

Reply:
I guess I'm wondering why the spawner doesn't populate it

Reply:
like automatically populate it? I am not sure. Maybe by default, they don't want to overwrite the os.env. There maybe a flag setting I haven't figured out yet.

Reply:
> Maybe by default, they don't want to overwrite the os.env.

I'm confused now - I thought self.environment was separate from os.env. It seems like you're overwriting self.environment with os.env. My basic assumption is that if that was the right thing to do the Jupyter devs would've done it, but maybe I'm wrong.

Reply:
yea. That's exactly what I am trying to do here, copy missing os.env to user's spawner env. It has things like `YARN_CONF_DIR` and `SPARK_MASTER_URL` that we need. But it's not automatically set to the user's environment if I don't explicitly set them.

Reply:
Ok. I would probably use an allow list here rather than dumping everything in there but I leave that up to you

Reply:
👍 I will dump everything for now just in case I might miss something.